tcp_conn.go

```diff
@@ -0,0 +1,39 @@
+package tcp
+
+import (
+	"errors"
+	"net"
+	"syscall"
+	"unsafe"
+
+	"github.com/mikioh/tcp"
+)
+
+// This is only needed because mikioh/tcp doesn't accept wrapped connections.
+// See https://github.com/mikioh/tcp/pull/2.
+
+type tcpConn struct {
+	net.Conn
+	c syscall.RawConn
+}
+
+// newTCPConn returns a new end point.
+func newTCPConn(c net.Conn) (*tcp.Conn, error) {
+	type tcpConnI interface {
+		SyscallConn() (syscall.RawConn, error)
+		SetLinger(int) error
+	}
+	var _ tcpConnI = &net.TCPConn{}
+	cc := &tcpConn{Conn: c}
+	switch c := c.(type) {
+	case tcpConnI:
+		var err error
+		cc.c, err = c.SyscallConn()
+		if err != nil {
+			return nil, err
+		}
+		return (*tcp.Conn)(unsafe.Pointer(cc)), nil
```

Comment:
This is 🤢🤢🤢.

The problem here is that `mikioh/tcp.NewConn` only accepts `net.TCPConn`s, but not wrapped TCP connections.
I submitted a rather trivial fix in https://github.com/mikioh/tcp/pull/2 a while back, but it looks like the repository is not maintained any more.

Reply:
Are we assuming the layout of the conn returned by multiaddr/net? That's really not something we can do.

Reply:
That's not what we're doing. We're assuming here that the layout of `tcpConn`, defined a few lines above, is the same as the layout as the [`tcp.Conn`](https://github.com/mikioh/tcp/blob/master/rawconn.go#L20-L23) defined in the tcp package.
Unfortunately, although the `tcp.Conn` is public, it has a private member which we need to initialize.